### ISSUE TITLE
Disable type checking for node_modules entirely

### ISSUE THREAD

**Issue by gempir** [microsoft/TypeScript#40426]
## Search Terms

skipLibCheck
node_modules
ignore library
exclude

## Suggestion

Either a new option or the existing option `skipLibCheck` should be able to disable type checking for node_modules.

## Use Cases

A library author might have a faulty import or might only provide typescript files for his library and not transpiled code. As a user of this library I want to be able to disable type checking for a library (or all) because I trust this library has been tested enough in other ways. 

## Examples

I had the case that a library came with a `dist/` folder and almost all types were perfectly arranged in .d.ts files. But one file was making an import from `../index.d.ts`(which seems to be for development purposes, the lib main file is `dist/index.d.ts`) which in turn then imported from `src/`

And now in src the actual typescript files got imported and got type checked against my tsconfig.json

The library author used a different tsconfig and didn't use strict type checking, but I do and that's why the type check fails for me.

## Checklist

My suggestion meets these guidelines (in case of a new option):

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by DanielRosenwasser** [comment]
> As a user of this library I want to be able to disable type checking for a library (or all) because I trust this library has been tested enough in other ways.

Okay, this is one of the reasons you might use `skipLibCheck`.

>I had the case that a library came with a `dist/` folder and almost all types were perfectly arranged in .d.ts files. But one file was making an import from `../index.d.ts`(which seems to be for development purposes, the lib main file is `dist/index.d.ts`) which in turn then imported `src/ClassA.ts`

This is the opposite of the scenario you're describing. The library author has shipped incoherent types, and who's to say what the correct interpretation of that is? It is **good** that TypeScript errors here, even if it is an overall terrible user experience.

____

"Disabling type-checking" can mean a lot of things. Having the package come in as `any` is one possibility. If that's what you're looking for, then you can add a single global file with `declare module "incorrectly-authored-library";` until the package ships correct types. Otherwise, I'm not sure what you might have in mind.

**Comment by DanielRosenwasser** [comment]
I want to mention after reading this a second time: shipping `.ts` files is generally nice when using `--declarationMaps` but I don't know how to best address the issue of different settings between your project and the library's project.

**Comment by gempir** [comment]
Yeah there are lots of good reasons to actually have the sources of a library available, I like to explore them in my IDE sometimes or maybe while debugging etc. 

But I don't actually want to have them checked by the typescript compiler, hence the skipLibCheck. It's just so confusing to me why the skipLibCheck is only intended for typing files and not for `.ts` files

**Comment by RyanCavanaugh** [comment]
This scenario is going to be completely busted for a lot of reasons. It seems like the option one would really want, if they found themselves stuck in it, would be to say that `.ts` files under some given paths should be treated as if they were `.d.ts` files (don't emit, don't typecheck under `SLC`, don't use to compute `rootDir`, etc) but have their implementations used for inference. That said, this is just critically a *misconfiguration* and it's possibly counterproductive to do gymnastics to make it work.

**Comment by gempir** [comment]
I think `.ts` files imported by `.d.ts` files should be treated the same way, so if skipLibCheck is enabled it should ignore them. 

If you are importing `.ts` files from actual source files like `dist/main.js` importing `../src/ClassB.ts` then i just expect it to break. That shouldn't be ignored.

**Comment by shadow-light** [comment]
Could there be a way to just silence errors from certain paths?

I don't want to make a whole import from node_modules type `any` as 99% of the types are fine and are useful for code completion, but I don't want to hear about the 1% of types that don't adhere to my tsconfig when I can't do anything about them (aside from fork the whole module for the sake of a couple of errors).

Noting that a single type check fail would break CI unless I silence it some how.

**Comment by shadow-light** [comment]
(I tried `@ts-ignore` above the import with the issue, but the errors are with certain files in that imported module, so obviously didn't work)

**Comment by NonkelDaniel** [comment]
I am having a very similar issue with a node_module that is included locally in package.json with a path:  `file:...`. there are type errors happening inside of that library and the compiler does not ignore them despite having `"skipLibCheck": true` in the config.

**Comment by shadow-light** [comment]
> I had the case that a library came with a dist/ folder and almost all types were perfectly arranged in .d.ts files. But one file was making an import from ../index.d.ts(which seems to be for development purposes, the lib main file is dist/index.d.ts) which in turn then imported src/ClassA.ts

> The library author used a different tsconfig and didn't use strict type checking, but I do and that's why the type check fails for me, because now the typescript source files of the library author are type checked.

Just ran into this exact scenario trying to type check my build scripts. Seems like not many people do that so some build-related modules have issues that go unnoticed.

**Comment by shadow-light** [comment]
Here's an example of a project that seems to be doing something benign, simply importing an interface for the sake of type checking. But the import path points to distributed typescript source files rather than the compiled JS. And while there aren't any issues with the types of that module, because my project has stricter types it causes issues.

https://github.com/underfin/vite-plugin-vue2/pull/125/files

**Comment by nitram-mural** [comment]
I'm having this problem with the lib **tst-reflect**/**tst-reflect-transformer**: "TS6133 declared but value never read"
I want that strict checking on our project but would love if it was not enforced on specific libraries I hand pick.

Compiler opts, ran with ttypescript's `ttsc`:
`  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "module": "commonjs",
    "noEmitOnError": true,
    "noImplicitAny": true,
    "noImplicitReturns": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "sourceMap": true,
    "strict": true,
    "strictFunctionTypes": false,
    "useUnknownInCatchVariables": false,
    "noImplicitThis": false,
    "target": "ES6",
    "lib": [
      "es6",
      "dom",
      "es2015.promise",
      "es2016.array.include"
    ],
    "outDir": "build",
    "plugins": [
      { "transform": "tst-reflect-transformer" }
    ]
  },`

**Comment by icywolfy** [comment]
We have been forced to disable `strict: true` on all our repos.  
Third-party libraries ship only `.ts` files.  The `package.json` has entry point as `index.ts`.  After a year of requests and pressure, no progress has been made convincing other companies to make changes to what they publish.

The code compiles properly, and works at runtime.  However, we wish to have `strict: true` enabled in our builds.  However this is not possible.  Required packages publish only raw `.ts` files, which do not meet `strict: true`, preventing automated enforcement of standards within code that we *do* have control over.  


Being unable to enable any strict typing, solely because we consume third-party packages that don't meet the same strictness, and of which we have no leverage for change: 
- use of `any`, 
- function type mismatches (`../../node_modules/@***/***/src/sdk/*****Api.ts:302:9 - error TS2322: Type '(value?: void | PromiseLike<void>) => void' is not assignable to type '(value?: unknown) => void'`
- `../../node_modules/@***/***/src/sdk/*****Api.ts:21:16 - error TS2532: Object is possibly 'undefined'`


We have a manual process that is sometimes applied: 
- check out a PR, 
- manually perform a `tsc --noEmit --project tsconfig-strict.json`,
- filter out hundreds of `errors` from `node_modules/` .ts file only libraries
- find errors in code we own
Which is error prone, and allows typing problems in our code to slip through.


I had suggested something similar to this : https://github.com/microsoft/TypeScript/issues/44205.  But it was rejected as "configuration issue".

**Comment by aprilmintacpineda** [comment]
Can we please get a flag to suppress any error messages from files originating from `node_modules`? Doesn't matter if there are 1 million errors if they are in `node_modules`, just ignore them, don't display them, and exit with code 0. The only time when we should exit with code 1 is when there are errors anywhere else other than `node_modules`

**Comment by jespertheend** [comment]
My project is fully in js but I do use IntelliSense with strict typing  so that I get early warnings when mistakes have been made. But since both `checkJs` and `strict` are enabled, I now get a ton of errors from node_modules.

Normally this wouldn't be a huge issue, but I have also enabled `Project Diagnostics`, which makes errors from all files show up, even if they're not open. Which makes it hard to see a warning whenever a file contains a mistake. If only a single file has an error, it immediately sticks out like a sore thumb, whereas if you permanently have a bunch of files and folders that are red, not so much. Not only that, the Problems view now has a huge list of errors that makes it impossible to find errors caused by me, unless I filter them first.

I also wish to include type checking in CI, but this makes that impossible without writing some sort of plugin to filter all unnecessary errors.

As far as I'm aware, there's a few options to fix this:
- Disable `checkJs` and place `@ts-check` at the top of all my files. This is prone to mistakes, but a linter rule might be able to prevent that.
- Add `@ts-nocheck` for all files with errors in node_modules. But it would have to be done with some sort of script to prevent overwrites when updating modules.
- Disable `Project Diagnostics`, but this means I won't get warnings from files that are not open. And making a change to a file doesn't necessarily mean new errors will occur in that file only.

But neither of these seem like a satisfying solution to me. If it were possible to add a list of paths to the tsconfig/jsconfig similar to the `include` and `exclude` properties, it would be possible to just ignore `"node_modules/**/*.js"` or `"node_modules/*"` entirely. This would basically function the same as placing `@ts-nocheck` at the top of each file. This seems like a nice solution to me, with the added benefit that files other than those in the `node_modules` directory can be configured to be ignored as well.

**Comment by snowinmars** [comment]
The irony is that this issue is not about no-name packages.

[Microsoft authentication library for js](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-core/src/error/AuthError.ts#L30) does throw an error during build under strict ts mode: `Type 'string | undefined' is not assignable to type 'string'`

Btw, check out [their average code quality](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-core/src/UserAgentApplication.ts#L1976)) Declared return type - 'Account', actual return type - 'Account | null'.

**Comment by RyanCavanaugh** [comment]
@snowinmars That library has a correct `typings` field so you shouldn't be seeing that error, can you clarify how your project is configured? https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-core/package.json#L24

An external library not having `strictNullChecks` on is not anyone's defect

**Comment by snowinmars** [comment]
> @snowinmars That library has a correct `typings` field so you shouldn't be seeing that error, can you clarify how your project is configured? https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-core/package.json#L24
> 
> An external library not having `strictNullChecks` on is not anyone's defect

Msal can be imported in two ways:
- you can import prebuild items from `lib-*` folders - and that will not break your build
- you can import sources from `src` folder itself to build it by yourself - and that will release the chaos

It's about `import { AuthError } from 'msal'` vs `import { AuthError } from 'msal/src/error/AuthError'`

Technically, one can say, that that's what prebuilds are for. But I would not agree: if developers doesn't want the source to be linked, they do not include it to npm package. So, it's a legal way to use a library, that just doesn't match with another's developer way.

If typescript compiler could solve this kind of issues, that'd be cool.

P.S. There is a `@azure/msal-browser` package with MUST better typings.

**Comment by RyanCavanaugh** [comment]
If the MSAL authors are telling you to import the TS sources directly from their packages, please tell me where so I can go tell them to fix their docs. This isn't a supported scenario and there's no end of problems this will cause.

**Comment by fatihaziz** [comment]
There is still no solution for “skipping node modules” typings? Because the package I use don't have correct typing on its own…

**Comment by bmarden** [comment]
Just ran into this myself. I'd really like to add a pipeline stage that runs a simple type check on our repo, but this is blocking that from happening. Is there really no workaround?

**Comment by nitram-mural** [comment]
I can't imagine the amount of packages that have this exact problem which end up just not using strict mode and introduce iffy code making everything worst downstream.

**Comment by hiddentao** [comment]
Having this issue myself as well :/

**Comment by cerskeenboy** [comment]
Same here.

**Comment by RyanCavanaugh** [comment]
Can people post "Here's how I ended up importing .ts files from node_modules" or "Here are the errors in a .d.ts in node_modules I had" instead of "same" ?

**Comment by shadow-light** [comment]
@RyanCavanaugh What's the difference between importing JS with typing files and import straight TS? I.e. what makes it difficult to support both?

I'm guessing it is to do with other TS features that are compiled to plain JS that aren't necessarily in the typing files?

These days every module has its own opinion about what level of JS to support, and an app may want to target an older version than a module wants to. I wonder if it would be better to publish plain TS and let app's decide what JS version to compile down to for both the app code and module code.

There's probably numerous challenges and performance issues with that, but just a thought...

**Comment by RyanCavanaugh** [comment]
There's no such thing as "importing straight TS" from another module which has a different configuration from yours (and, as you point out, different people having different configurations is very common). There are at least a dozen different compiler flags that might change the interpretation of a line of implementation code. The process of making `.d.ts` files takes away *most* of those such that any two projects can communicate in a sort of "common language", but if you take the `.ts` file itself then its true type behavior is not really known without also knowing its configuration.

We "could" support TS importing files and looking up their true configuration and reinterpreting the code differently in different regions of the code, but this is a *huge* mess with really bad performance and predictability characteristics. So all of this is to what end? The package can just publish its `.d.ts` and, under a modern module ecosystem, a `.js` file targeting a modern JS language variant, which you can bundle and downlevel to whatever target. Bundlers already do a great job of that, including features TS doesn't have like tree-shaking. We don't need to reimplement the wheel.

**Comment by nitram-mural** [comment]
> Can people post "Here's how I ended up importing .ts files from node_modules" or "Here are the errors in a .d.ts in node_modules I had" instead of "same" ?

Transformer to add reflection features to the compiler:
https://github.com/Hookyns/tst-reflect
https://www.npmjs.com/package/tst-reflect-transformer

**Comment by fatihaziz** [comment]
> > Can people post "Here's how I ended up importing .ts files from node_modules" or "Here are the errors in a .d.ts in node_modules I had" instead of "same" ?
> 
> Transformer to add reflection features to the compiler:
> https://github.com/Hookyns/tst-reflect
> https://www.npmjs.com/package/tst-reflect-transformer
> 
> 

🙏🙏🙏
Really appreciate you share this

**Comment by shadow-light** [comment]
Thanks @RyanCavanaugh, that's what I suspected. A future where you could just publish plain TS with config would be nice, but it isn't very realistic, as you point out.

Perhaps there could be a "How to publish a TS project" page in the docs? There are ones on publishing .d.ts files, but the average user may not even realise it's not appropriate to publish straight TS and just go ahead without bothering with generating .d.ts files.

**Comment by aspiers** [comment]
I'm not sure from the above comments whether disabling type checking is feasible or not.  But even if it's not, I didn't yet see anyone come up with a good reason why @shadow-light's [suggestion of simply silencing errors from `node_modules`](https://github.com/microsoft/TypeScript/issues/40426#issuecomment-901857106) wouldn't be a perfectly valid solution here.  Did I miss something?

Even nicer would be an extra option so you could choose between either silencing them or downgrading them from errors to warnings.

As @aprilmintacpineda [observed above](https://github.com/microsoft/TypeScript/issues/40426#issuecomment-1015376906), it's surely an extremely common use case to want to be able to run `tsc --noEmit` on your own code, and have it run checks as strictly as possible, whilst not complaining about third-party dependencies in `node_modules`.  If this use case is not possible then this feels to me like a serious omission.

I totally empathasise that from perspective of someone maintaining or working on TypeScript itself, the most desirable outcome is probably to harness the brokenness of the third-party dependency in order to provide the impetus for its consumers to exert pressure on the third party to fix their package.  And indeed in an ideal world that would provide the most benefit for the whole TypeScript ecosystem.  But as we can clearly see from the number of :+1: reactions to this report, as well as several other related feature requests like #30511, #44205 etc., unfortunately there are just too many cases of brokenness for that to be not pragmatic.  And when the only available alternatives are just disabling or weakening the checks on your own code, that approach just ends up hurting devs more than helping them.

IMHO it's really important to not let perfection become the enemy of the good here.

**Comment by nitram-mural** [comment]
Being common enough as mentioned, adding an entry for this scenario in the official documentation could add a lot of value. I mean running twice with two different configs to workaround isn't the first thing that comes to mind, usually.

One run for CI checks with no-emit and restricted fileset,
and the other to run proper compilation but discarding the exitcode in CI with `tsc ... ||true` or a Windows command counterpart.
Although for my case of a compiler transformer dependency I don't know how limiting the imported paths could be possible (I would love a doc page about that!)

**Comment by RyanCavanaugh** [comment]
"Simply silencing" is really dangerous.

For example, let's say you reference an npm module that works in the browser, but your configuration doesn't include the DOM. The module declares some overloaded function like
```
function doSomething(x: Window): void;
function doSomething(x: number): void;
```
Normally our behavior in the presence of an unresolved type is to treat it as `any` (**note**: treating this as `never` or some other never-assiganble-to-type creates an entirely different but equally-bad set of problems, so please don't just counterpropose that we do that instead). So effectively the function signature becomes
```
function doSomething(x: any): void;
function doSomething(x: number): void;
```
which you can call with `doSomething("hello world");` without error

What you'd really prefer here is an error stating that there is a configuration problem.

The same thing might happen if an npm module uses some built-in type from a newer version of TypeScript than the one you're using, or even newer syntax.

The vast majority of the time that you see an error from `node_modules`, you have a legitimate configuration problem that needs to be addressed before a coherent interpretation of your code can exist.

The implications of letting this run rampant are also very troubling. It's a *lot* slower to process `.ts` than `.d.ts` files. If everyone can just get away with publishing misconfigured `.ts` files instead of properly publishing `.d.ts` files, then over time the number of people publishing broken packages is going to invariably go up. *Everyone* is going to pay the performance tax of re-re-re-re-re-inferring the return types of these function bodies, forever, since this is a ratchet that only goes in one direction. To make matters worse, errors are *expensive* -- there is a lot of complex logic that we only run in error cases, since in general it's assumed that spending an additional 50ms of CPU time to compute a really good and accurate error is a good trade-off. If we're generating these errors and then just throwing them away later, it's going to waste a ton of time forever.

I realize it really seems like there are simple easy cost-free no-problems solutions here, but there really are not, apart from people not publishing misconfigured packages.

**Comment by aspiers** [comment]
I don't think there's any doubt that as a general principle silencing errors is dangerous; you're preaching to the choir in that regard ;-)

However, when the only errors found are clearly minor third-party issues which have no real impact on the *actual* safety of the code in question, it's *more* dangerous *not* to be able to silence them, because this means the only way developers can get a clean run in their CI is to drastically weaken the checks being done.

And we have seen plenty of evidence for developers in that position, both in this ticket and related ones.  People are resorting to desperate measures like [disabling strict](https://github.com/microsoft/TypeScript/issues/44205#issuecomment-998169708) checks [everywhere](https://github.com/microsoft/TypeScript/issues/44205#issuecomment-989206997).

> The vast majority of the time that you see an error from node_modules, you have a legitimate configuration problem that needs to be addressed before a coherent interpretation of your code can exist.

That's not been my experience.  It's often errors in parts of a third-party library which my code doesn't even use.  And "the vast majority" implicitly acknowledges that there are at least some cases where there isn't a legitimate configuration problem.  How can we deal with those cases?

I appreciate that there are counter-examples - probably many in fact.  But it is not helpful to generalise from "there are cases where silencing errors is a really bad idea" to "no one should ever be allowed to silence errors, ever".  Tools should work for developers, not against them.  If a tool allows a developer to shoot themselves in the foot, as long as it comes with sufficient warning, that's fine - it's up to that particular developer to be careful.  But please don't prevent responsible developers from solving problems in a pragmatic way in order to stop other developers being irresponsible.  If yarn adopted that approach with things like peer dependencies instead of just issuing helpful warnings, it would probably be totally unusable.

The documentation could provide plenty of warnings about the risks, and if you're still worried even that's not enough, it could simply turn the errors into non-fatal warnings.  That would allow CI to run but still continually raise awareness about the risks until they're fixed.

> The implications of letting this run rampant are also very troubling. It's a lot slower to process .ts than .d.ts files.

Maybe I'm missing understanding something, but isn't this mixing two separate things?

1. the ability to silence type checking for selected paths (e.g. `node_modules`, but ideally it could be way more specific than that, like a single badly-behaved third-party module)
2. doing type-checking on `*.ts` files in libraries vs. `*.d.ts`

I know the original post mentioned both, but the latter was just in some example use cases; it was not representing the entirety of the problem space.  The primary concern of this issue is with the former.

Let's say that I don't disagree with you about the latter, and just focus on cases where libraries are not publishing misconfigured `*.ts` files and are publishing `*.d.ts` files.  That still leaves the question of why it's not OK to be able to silence errors relating to `*.d.ts` files in selected paths (or at least downgrade to warnings), given that in at least some cases this is much more pragmatic and safe than the alternative of disabling important checks over the entire codebase.  What is the problem in this case?

**Comment by RyanCavanaugh** [comment]
> That still leaves the question of why it's not OK to be able to silence errors relating to *.d.ts files in selected paths

This is effectively already what `skipLibCheck` does, though - basically the only errors you'll see under this setting are the ones that are true problems that TS encountered when evaluating the relevant portions of the .d.ts for your own code.

**Comment by RyanCavanaugh** [comment]
> Maybe I'm missing understanding something, but isn't this mixing two separate things?

OP was specifically referring to importing `.ts` files:

> A library author might have a faulty import or might only provide typescript files for his library and not transpiled code. As a user of this library I want to be able to disable type checking for a library (or all) because I trust this library has been tested enough in other ways.

and

> I understand that libraries shouldn't publish *.ts files, but sometimes, they just do, and we can't control every other library easily.

**Comment by aspiers** [comment]
> > That still leaves the question of why it's not OK to be able to silence errors relating to *.d.ts files in selected paths
> 
> This is effectively already what `skipLibCheck` does, though

I already tried that and in several cases it didn't help.  Your use of the word "effectively" seems to acknowledge that there are caveats, and so does [your comment on #47387](https://github.com/microsoft/TypeScript/issues/47387#issuecomment-1010123047).

> basically the only errors you'll see under this setting are the ones that are true problems that TS encountered when evaluating the relevant portions of the .d.ts for your own code.

I think maybe we have slightly different ideas of what constitutes "true problems".  You seem to be viewing it from a more purist perspective of whether there are genuine compilation issues within the codebase.  This is of course perfectly valid and quite understandable given the nature of your work on TypeScript!

In contrast however, I am more concerned with being able to implement needed functionality in my codebase whilst having its CI and linting set up for a pragmatic balance between theoretical correctness and being able to get my job done.  That means that if I encounter some errors which I *know* are not serious issues *in my context* (for example if they relate to library code which I'm not even using), then as an experienced software engineer I want to be able to make a judgement call on the risk, temporarily remove those obstructions by downgrading them to warnings, and then return to them later as lower priority concerns.

I suspect many of the people following this issue and similar ones like #44205, #30511, #47387 are closer to my position than yours, otherwise we wouldn't be seeing so many :+1: and :heart: emojis on the OP.

> > Maybe I'm missing understanding something, but isn't this mixing two separate things?
> 
> OP was specifically referring to importing `.ts` files:
> 
> > A library author might have a faulty import or might only provide typescript files for his library and not transpiled code. As a user of this library I want to be able to disable type checking for a library (or all) because I trust this library has been tested enough in other ways.

Specifically but *not exclusively*.  Notice the repeated use of the word "might".

> A library author *might* [...] or *might* [...]

Two examples were given, the first of which had nothing to do with libraries providing typescript source files.  And this was under a "Use cases" header, after the original issue title and "Suggestion" header, neither of which mentioned anything about libraries providing them.

But regardless of the OP's intentions, my point is that there definitely two orthogonal topics here, and we need to consider the ability to selectively silence or downgrade errors from selected paths, **regardless** of whether those errors are being caused by `*.ts` files in third party libraries.  If you prefer, I'm happy to submit a new issue to keep them separate, but I don't think the title and topic of conversation would change significantly.

**Comment by MicahZoltu** [comment]
Please redirect me if this isn't the right thread to bring this up; searching the issues on this topic is hard because `ts`, `d.ts`, `declaration` show up in almost every issue in this repository.

I have a dependency with an `index.js`, `index.d.ts`, `index.d.ts.map`, and `index.ts` all in the root folder.  `package.json` for the dependency looks something ilke:
```json
{
  "type": "module",
  "main": "index.js",
  "module": "index.js",
  "files": [
    "index.js",
    "index.d.ts",
    "index.d.ts.map",
    "index.ts"
  ]
}
```
For reasons I cannot understand, when I import it as a named package (e.g., `import { ... } from 'package'`) TSC finds `index.ts` and tries to use that instead of using `index.d.ts`.

I can work around it temporarily by manually editing the dependency's `package.json` to contain `"types": "index.d.ts"`, but it really feels like that should be the default, not `index.ts`.

Is this something already being talked about above?  It seems like if TSC just defaulted to `.d.ts` when there is no `types:` field in package.json then some subset of problems would go away.  Why is it preferring `.ts` files for named packages?

**Comment by JacobDel** [comment]
@avidj the only option I see is to fork the library and generate the *.d.ts files yourself.
Or at least that's what I'm trying to do in my case

**Comment by erikeckhardt** [comment]
I turned on `skipLibCheck: false` recentl;y precisely because I *want* type checking on my hand-written .d.ts files. These do not duplicate anything in `.ts` files. They contain things like ambient interfaces that are implemented in concrete classes elsewhere in .ts files. Or special types that are used throughout my project.

First, I had several errors in my .d.ts files that I *wanted* to know about. This was great! I regretted that checking .d.ts files is not turned on in TypeScript by default. This seems weird. Why wouldn't I wan that?

But then I discovered the node_modules issue. And as has been repeated over and over in this discussion, there is not a single thing I can realistically do about type errors in third-party packages. For example, there are errors about Dom stuff sinceI'm writing a pure back-end Node.js library module and do not want DOM stuff. My `setTimout` returns a `NodeJS.Timeout`, not a `number`. That's fine with me. I can't control whether authors of sub-sub-sub-sub-sub-dependencies wrote code for both Node.js and the browser. I just DO NOT CARE about these errors.

The argument that it is dangerous to allow developers to exclude `node_modules` from type checks when `skipLibCheck` is false makes no sense at all. Not allowing me to do this makes my own code more likely to be broken!

I NEED to be able to type-check my own project's .d.ts files without getting errors about problems in the `node_modules` folders somewhere-or-other. Otherwise, I simply cannot use `skipLibCheck: false` which will provide less type safety.

To any TypeScript contributor resistant to giving we developers what we're asking for here, I seriously have to say "shame on you." Stop making it difficult for me to get better quality in my own projects. Stop it. Please. Just stop.

I have work to do. I can't fix the world. I can't put pressure on these third-party packages. I'll lose my job or my contract if I don't deliver. I simply don't have time or energy to help make the world perfect. I learned to program almost 4 decades ago. I like sharp knives. Please give me sharp knives. I know how to use them. I have the scars to prove it, and I don't cut myself much any more. And if I *understand* a system and can control it, I can heal from the cuts quickly, anyway.

While I wouldn't _like_ it, I'd be content enough if I had to silence .d.ts errors from `node_modules` on a project-by-project or even a file-by-file basis in some config somewhere. I know what I'm doing. Now let me do it.

I most humbly and abasedly beg you all on bended knee and with bowed head to freaking help me get my job done.

**Comment by danielschnetler** [comment]
[Look at this comment](https://github.com/microsoft/TypeScript/issues/47387#issuecomment-1421563426)

**Comment by appden** [comment]
I also strongly believe the most pragmatic path forward would be the skip type checking files under a `node_modules` directory (*at least* when `skipLibCheck` is enabled). I haven't seen a compelling argument in favor of continuing with existing behavior that outweighs the pain this causes developers who cannot control third-party libraries.

The workaround I've found for this annoyance is to leverage [project references](https://www.typescriptlang.org/docs/handbook/project-references.html) to prevent type checking certain problematic dependencies. The `skipTypeChecking` utility function  ([here](https://github.com/microsoft/TypeScript/blob/v4.9.5/src/compiler/utilities.ts#L7233)) returns `true` for files that are a part of a project reference.

**Comment by MicahZoltu** [comment]
I'm speculating a bit here, but I would *guess* that "skip node_modules" would be non-trivial to implement.  The compiler has to look at type definitions in node_modules for any dependency you reference, so it has to do *some* type checking.

I suspect the best that could be done is something like "ignore warnings in node_modules" or perhaps "ignore non-critiral errors in node_modules".  This is a bit risky though because it means the compiler would need to make some reasonably guesses about what types it should return in these situations.

**Comment by lillallol** [comment]
So I use TypeScript type checking without the need to compile, i.e. I am manually writing all my types in `.d.ts` files, and import-type-annotate values in `.js` via `/**@type {import("./path/to/file.js").IMyType}*/`. Type checking `.d.ts` files (both outside and inside `node_modules`) is disabled when `skipLibCheck:true`, so I am forced to disable it.

For a project of mine that I use `"noUncheckedIndexedAccess": true`, and has a dependency another project of mine that does not use `"noUncheckedIndexedAccess": true`, I start getting linting errors when `tsc --noEmit`. That is because I also use `"maxNodeModuleJsDepth": 8`, because I want to make browser compatible imports so that I do not have to bundle during the development stage.

Here is what I did to run `tsc --noEmit` and ignore all the errors outside `node_modules` and project root:

```js
import { ts } from "../../adapter/typescript.js"
import { fs } from "../../adapter/fs.js";
import { path } from "../../adapter/path.js";

export const runTscInProjectRootOutsideNodeModules = () => {
    const absolutePathToProjectRoot = path.resolve("./");
    const absplutePathToNodeModules = path.resolve("./node_modules");
    const absolutePathToTsConfig    = path.resolve("./tsconfig.json");
    const parseJsonResult = ts.parseConfigFileTextToJson(
        absolutePathToTsConfig,
        fs.readFileSync(absolutePathToTsConfig, { encoding: "utf8" })
    );
    
    const tsConfig = ts.parseJsonConfigFileContent(
        parseJsonResult.config,
        ts.sys,
        path.resolve("./")
    );
    
    const program = ts.createProgram({
        rootNames : tsConfig.fileNames,
        options   : tsConfig.options,
    });
    
    const diagnostics = ts.getPreEmitDiagnostics(program);
    const diagnosticsInProjectRootOutsideNodeModules = diagnostics.flatMap((diagnostics) => {
        const {file} = diagnostics;
        if (file === undefined) throw Error(); //@TODO I have no clue what this case is for
        const absolutePathToCurrentFile = path.resolve(file.fileName);
        if (
             absolutePathToCurrentFile.startsWith(absolutePathToProjectRoot) &&
            !absolutePathToCurrentFile.startsWith(absplutePathToNodeModules)
        ) return [diagnostics]
        return [];
    })
    diagnosticsInProjectRootOutsideNodeModules.forEach(({
        file: sourceFile,
        start,
        code,
    }) => {
        if (sourceFile === undefined) throw Error();//@TODO I have no clue what this case is for
        if (start === undefined) throw Error();//@TODO I have no clue what this case is for
        
        const fileName = sourceFile.fileName;
        const relativePath = path.relative(absolutePathToProjectRoot,fileName);
        const {character,line} = sourceFile.getLineAndCharacterOfPosition(start);
        console.log(`${relativePath}:${line + 1}:${character + 1}`,`TS${code}`);
    })
    if (diagnosticsInProjectRootOutsideNodeModules.length > 0) process.exit(-1);
}
```

**Comment by dejayc** [comment]
> there is not a single thing I can realistically do about type errors in third-party packages

I mean, you could always switch to a programming language that's not held up by ideological fantasies that take half a decade to resolve, on average 🤣

If it's true that you learned to program 40 years ago, you'll be well retired by the time TypeScript addresses its most serious flaws.

**Comment by RyanCavanaugh** [comment]
@dejayc be constructive.

**Comment by hudson155** [comment]
I'm a bit confused here.

If there are `.ts` files in any of my dependencies (such as `@auth0/auth0-spa-js`), those files are required to conform to the same `compilerOptions` that my own sources files are? Auth0 does not conform to `noImplicitReturns`, for example, so it seems like I'm forced to resort to the lowest common denominator and disable compiler options that I want to enforce but some of my dependencies don't conform to.

Am I missing something here?

**Comment by ritik3236** [comment]
any solution?
https://github.com/microsoft/TypeScript/issues/40426#issuecomment-1015376906

**Comment by abumalick** [comment]
> ~I'm a bit confused here.~
> 
> ~If there are `.ts` files in any of my dependencies (such as `@auth0/auth0-spa-js`), those files are required to conform to the same `compilerOptions` that my own sources files are? Auth0 does not conform to `noImplicitReturns`, for example, so it seems like I'm forced to resort to the lowest common denominator and disable compiler options that I want to enforce but some of my dependencies don't conform to.~
> 
> ~Am I missing something here?~
> 
> As luck would have it, despite debugging for hours prior, I figured out my issue 10 seconds after posting this comment. I turns out I was indeed missing something here. I was importing a TS file when I should've been importing a JS file.
> 
> Replacing
> 
> ```ts
> import { AuthorizationParams } from '@auth0/auth0-spa-js/src/global';
> ```
> 
> with
> 
> ```ts
> import { AuthorizationParams } from '@auth0/auth0-spa-js';
> ```
> 
> solved my problem! 🎉
> 
> Feel free to delete this comment — I just left it in case it helps a fellow straggler.

Thank you very much @hudson155 , this fixed my problem.

**Comment by ValentinGurkov** [comment]
I have a monorepo.  Similarly to other people here, I wanted to implement a type check step and have run into this issue.
My monorepo package B imports package A as a workspace dependency (pnpm creates symlinks here).
When type checking package B it follows through to package A, which is something I don't want, package A has its own type check step with its own tsconfig.

**Comment by akwodkiewicz** [comment]
>I have a monorepo. Similarly to other people here, I wanted to implement a type check step and have run into this issue.
My monorepo package B imports package A as a workspace dependency (pnpm creates symlinks here).
When type checking package B it follows through to package A, which is something I don't want, package A has its own type check step with its own tsconfig.

Adding my 2 cents, because I have the same issue with a monorepo and wanted to add more context.

My current workflow in a monorepo is to run `tsc` on a library just to emit the `d.ts` files, so that a different project inside that same monorepo can use those, in the same way as if it was a package downloaded from npm. It works, but it requires running the `tsc`, even though I have the types already there, a folder away.

I thought I'll instruct TS to use the library's source code instead and skip the compilation step completely, making the feedback loop instant. So I first made sure my projects are now relying on Node16 moduleResolution, and then used the `exports` entry in the library's manifest, to point `"types"` to `"./*": "./src/*.ts"`.

But because the library was written some time ago, with not-that-strict rules, the app that's using it is now throwing compilation errors, because app's tsconfig is stricter than the library's.

I now have to either:
1. invest time in making the old library stricter
2. make the new app less strict
3. still rely on additional step of generating `d.ts` files with the help of `tsc`, before working with the app -- keep in mind that the additional compilation is not only necessary when writing the code for the new app, but also when running other type-aware tools, such as ESLint

And the whole dilemma actually started when I decided to move out of cross-project path aliases (which are misused by people working on monorepos) and rely on proper package imports via Yarn.

**Comment by devdumpling** [comment]
Just wanted to +1 the above comments. 

Not sure if perhaps it deserves a separate issue thread entirely, but like @akwodkiewicz, I have a monorepo (pnpm, turborepo, typescript, Next.js). 

Currently I build all of those packages, as well as emit `.d.ts` for each, and consume them within an app. I have a typecheck step for each of these packages which has worked fine since each package has a `types` entrypoint pointing to the declaration file generated at build. 

However, I no longer actually need to build these packages, as the app is fully capable of transpiling them. Furthermore it turns out the app transpiling is quite a bit faster than running build in each package. This is great, big win. Problem is, now I have no way of typechecking my packages individually (as they all use workspace links and each of them would now point to a `.ts` or `.tsx` file instead of the no-longer-generated `.d.ts`). 

Like others above, if I try to typecheck a single package, it will end up typechecking the dependencies of that package as well, which is both redundant (since I typecheck these in separate processes) and potentially buggy, as the `tsconfig` for various packages may differ in strictness.

I'd love to be able to simply exclude or skip, but that doesn't seem to work (e.g. `exclude: "@scope/**`) or something. 

Are there any other suggestions for monorepo/workspace users running into this issue?

**Comment by kumikumi** [comment]
I have a frontend and a backend both written in TypeScript. The backend runs on Bun, and frontend is built with Vite. Using the `import type` syntax, I'm able to import some types from backend to my frontend code. I have made the decision to not generate or maintain separate `.d.ts` files, instead I'm just relying on type inference to get the types of values that the backend's API returns to frontend.

This works.

My only remaining problem is that the frontend's build process is now spitting out errors from the backend. It's obviously missing all of the backend libraries, types for Bun etc. In my case, none of it matters, as none of that code is going to be included in the front-end build anyway (because `import type` lines get erased when emitting code).

Please allow an option to either

A) only show errors from within the project, ie. from files included in the `include` section of tsconfig.json

(which would make a sensible default IMO, but it's debatable I guess)

or

B) allow specifying an `excludeTypechecking` section in tsconfig.json, where you can specify a list of globs to be excluded from typechecking. (OP could specify `node_modules`, I could specify my backend directory to be excluded in the frontend's tsconfig.json)

I disagree with the notion that the above is "really dangerous". It is no more dangerous than using //@ts-ignore or //@ts-nocheck, which are options that we are trusted with.

TypeScript is a tool - please let us apply it selectively.

**Comment by re-gor** [comment]
> I understand that libraries shouldn't publish *.ts files, but sometimes, they just do, and we can't control every other library easily.
> I think typescript should not try to apply this kind of setting (here noImplicitAny) to the node_modules simply because it is meant to be applied to the user code being compiled, not the external dependencies.

For those who are still struggling, especially if you live in a monorepo (like Lerna or Pnpm), my colleague has found a nice workaround

Putting in `"moduleSuffixes": [".d", ""],` in `tsconfig.json` will force Typescript to use `.d.ts` files in first place instead of `.ts`. 

Helped me when Typescript tried to compile a neighboring project

**Comment by akwodkiewicz** [comment]
> Putting in "moduleSuffixes": [".d", ""], in tsconfig.json will force Typescript to use .d.ts files in first place instead of .ts.

That solves part of the problem: (if what you say it's true 🤞 ) with this particular `moduleSuffixes` value you can now generate declaration files for the less-strict project and work on the dependant stricter project without compiler errors.

But you are still forced to first generate declaration files for that less strict project. The best way (from DX perspective) would be to just rely on the existing `.ts` files without additional build steps.

**Comment by russellmcc** [comment]
`bun` currently encourages publishing `.ts` source files directly to npm on [this page](https://bun.sh/docs/runtime/modules) (search for "Shipping TypeScript").  It sounds like this is explicitly _unsupported_ by `tsc`, even with the "bundler" module resolution.  In particular, as in this issue, it's impossible to skip typechecking for imported files, even if they are from a different package. This makes it impossible for a package to use stricter settings than its dependencies.  Is this right? I'm not sure what it would take to support this on the `tsc` side, but as a developer it's certainly convenient to publish `ts` files directly without having to rely on a bundling step.

**Comment by borisrakovan** [comment]
I wrote this workaround script that solves this without needing to relax the TypeScript configuration of your project. It adds a @ts-nocheck directive to the top of all problematic files of the 3rd party libraries, essentially silencing the errors when your project is compiled.

You can run it automatically on install by adding this to the `scripts` section in `package.json`

`"postinstall": "tsx scripts/fix-spl-token-types.ts"`

```typescript
import { readFile, writeFile } from 'fs/promises';
import { join } from 'path';

type PackageConfig = {
	// Package name as it appears in node_modules
	name: string;
	// Base directory within the package where source files are located
	sourceDir: string;
	// List of problematic file paths relative to sourceDir
	problemPaths: string[];
};
const PROBLEMATIC_PACKAGES: PackageConfig[] = [
	{
		name: '@solana/spl-token',
		sourceDir: 'src',
		problemPaths: [
			'extensions/tokenGroup/state.ts',
			'extensions/transferFee/instructions.ts',
			'extensions/transferHook/instructions.ts',
			'extensions/transferHook/seeds.ts',
			'extensions/transferHook/state.ts',
			'state/account.ts',
		],
	},
];

async function addTsNoCheck() {
	try {
		// Go up to common's parent (monorepo root)
		const rootNodeModules = join(__dirname, '..', 'node_modules');

		for (const pkg of PROBLEMATIC_PACKAGES) {
			const packagePath = join(rootNodeModules, pkg.name, pkg.sourceDir);
			console.log(`Processing ${pkg.name} in ${packagePath}`);

			for (const problemPath of pkg.problemPaths) {
				const fullPath = join(packagePath, problemPath);

				try {
					const content = await readFile(fullPath, 'utf-8');

					if (!content.includes('// @ts-nocheck')) {
						const newContent = `// @ts-nocheck\n${content}`;
						await writeFile(fullPath, newContent);
						console.log(`Added @ts-nocheck to ${fullPath}`);
					} else {
						console.log(
							`@ts-nocheck already present in ${fullPath}`
						);
					}
				} catch (err: any) {
					if (err.code === 'ENOENT') {
						console.warn(`File not found: ${fullPath}`);
					} else {
						console.warn(`Could not process ${fullPath}:`, err);
					}
				}
			}
		}

		console.log('Successfully processed all packages');
	} catch (err) {
		console.error('Error processing files:', err);
		process.exit(1);
	}
}

void addTsNoCheck();
```

**Comment by natew** [comment]
Can't use verbatimModuleSyntax atm if even one package has a bad setup import (in my case expo-modules-core).

**Comment by ibrust** [comment]
I don't understand what the argument would be for not allowing the programmer to prevent type checking of node_modules. Honestly it is completely obvious that this should be supported, you already have global configs for this such as --skipLibCheck - if it were such a design problem why do you provide the global flags? 
Amazing

**Comment by hematy61** [comment]
> Can't use verbatimModuleSyntax atm if even one package has a bad setup import

I have the same issue. As soon as I turned on verbatimModuleSyntax, it started type checking dependencies.

**Comment by robpalme** [comment]
@hematy61 Probably you have an unused import that was previously being trimmed via the default "Import Elision" behavior. VMS turns that off.

If you trim it manually (delete the line, delete the binding, or mark the import/binding with `type`) then you should restore the original scope of the checker.

After trying this, if you still find that VMS increases the scope of the checker, please provide a minal repro. Because I think that would be new information and potentially a bug.

**Comment by Hookyns** [comment]
@RyanCavanaugh 
> Can people post "Here's how I ended up importing .ts files from node_modules" or "Here are the errors in a .d.ts in node_modules I had" instead of "same" ?


This is a real and easily reproducible problem. You're not intentionally importing TypeScript - you're simply installing a package that includes a .ts file listed under the `types` field in `package.json`. There are packages out there doing this, and the number is growing due to support for runtimes like Deno and Bun.

Consider a package with the following `package.json`:

> package.json
```json5
{
    "name": "incriminated-package",
    "types": "index.ts", // <==== this !!
    "module": "dist/index.js",
    "type": "module",
    "files": [
        "dist/index.js",
        "index.ts"
    ]
}
```

If that linked .ts file is less strict than your own codebase, you're in trouble. Why did someone publish a .ts file alongside the compiled JS instead of a .d.ts file? There are multiple reasons:
- To support Deno, Bun, and other runtimes that natively use .ts.
- It's sometimes impossible to generate .d.ts files due to limitations in certain bundlers. You might need specific plugins, and they don’t always work - especially with complex setups involving multiple exports, or when the folder structure of dist/ differs significantly from src/.


DEMO [here](https://github.com/Hookyns/ts-libcheck-issue), live example [here](https://stackblitz.com/github/Hookyns/ts-libcheck-issue) on stackblitz.

Error:
```
node_modules/incriminated-package/index.ts:2:5 - error TS2322: Type 'null' is not assignable to type 'string'.

2     property: string = null;
      ~~~~~~~~


Found 1 error in node_modules/incriminated-package/index.ts:2
```

The TypeScript file is not used during compilation or at runtime - only the JavaScript is. I mean, that TS has no real impact on project but typechecking. So why does this error matter when skipLibCheck is enabled? TypeScript doesn’t use that code; it doesn’t bundle it or execute it in any way.

Why can't it be ignored the same way .d.ts files are? If there's a valid scenario where it does matter, then maybe the solution is to introduce a separate config option - something like skipLibTsCheck?! It will not break anything and quite a lot of people want this for 5 years!

**Comment by somebody1234** [comment]
another example of `.ts` files (advertently or not) being included in the `types` field which will cause errors in stricter codebases.
- https://github.com/adobe/react-spectrum/pull/8294

(no hate to the authors! it's a good library)

**Comment by phil294** [comment]
I'd also welcome it if we could ignore node_modules errors per se, and/or have `skipLibCheck` also skip non-node_modules `.d.ts` files is pretty annoying. But I also stumbled upon something else:

I tried the suggested strategy outlined by Daniel, `declare module 'whatever'` to overwrite wrong typinges, but that doesn't if you're

1. using JS, not TS, but strict type checking
2. using `jsconfig.json`, not `tsconfig.json`
3. pass the `jsconfig.json` to `tsc` using `-p` option
4. the library ships as js files only

Example: `npm install vue-virtual-scroller`, import it in a `.js` file, set `strict: true` in `jsconfig.json`, run `npx tsc --noEmit -p jsconfig.json` and see errors that are *impossible* to get rid off, regardless of any extra `.d.ts` files you put in the directory.

I don't really know what's going on and the type check output seems also to be pretty different with jsconfig, but the solution is simple:

Rename `jsconfig.json` to `tsconfig.json`.

Maybe it's a bug. But my past experience in reporting bugs has been that everything is always working as intended and it's my misunderstanding at fault 😄. On an unrelated note though, it's so refreshing to see core TS contributors on GH issues debating with the community over and over and keep explaining and arguing even after years. This must be exhausting AF, but it's one of my favorite aspects about the language.

**Comment by repulsio** [comment]
Any solution here? I just want to simply exclude a folder from types

**Comment by onyedikachi23** [comment]
It's almost 5 yrs now. @RyanCavanaugh Please 🙏 add the feature.

**Comment by iva2k** [comment]
It is clear that without taking types from node_modules, all packages that have been used and provide return values into user code will be "disconnected" and lead to no type checking on the function calls, on returned types and derived values, which can be a lot. 

So it makes perfect sense to keep reading types from node_modules.

What does not make sense is to bark at every type issue in packages in node_modules, which could be a huge pollutant of the logs.

I sense that everyone coming to this issue simply wants to silence all the warnings and (maybe optionally) errors from node_modules, as they simply create a ton of noise, but are not actionable and not fixable (in a sense that 3rd party packages are not under user control).

What if there were an option "silence node_module type check messages" with 3 choices - show all (errors and warnings), show only errors, and hide/ignore all errors and warnings. Regardless of that setting, types that can be used from the node_modules packages should propagate through the type checking process. The ones that are broken or not useable, will be quietly ignored if setting asks for it. Default, if not specified, of course, should be like it is currently behaving - report all issues, so there are no surprises.

**Comment by onyedikachi23** [comment]
> It is clear that without taking types from node_modules, all packages that have been used and provide return values into user code will be "disconnected" and lead to no type checking on the function calls, on returned types and derived values, which can be a lot. 
> 
> So it makes perfect sense to keep reading types from node_modules.
> 
> What does not make sense is to bark at every type issue in packages in node_modules, which could be a huge pollutant of the logs.
> 
> I sense that everyone coming to this issue simply wants to silence all the warnings and (maybe optionally) errors from node_modules, as they simply create a ton of noise, but are not actionable and not fixable (in a sense that 3rd party packages are not under user control).
> 
> What if there were an option "silence node_module type check messages" with 3 choices - show all (errors and warnings), show only errors, and hide/ignore all errors and warnings. Regardless of that setting, types that can be used from the node_modules packages should propagate through the type checking process. The ones that are broken or not useable, will be quietly ignored if setting asks for it. Default, if not specified, of course, should be like it is currently behaving - report all issues, so there are no surprises.

This is exactly what we need. Good reasoning.

**Comment by sparr** [comment]
Today I encountered the behavior that tsc insists in doing error checking for node_modules that publish .ts sources, even with node_modules excluded and skipLibCheck set. I would like a way to disable that behavior.

**Comment by Pentadome** [comment]
I needed to import a .ts file from a node module and now i get typescript errors from that file. Why can't we set paths where we don't want type checking?

**Comment by everett1992** [comment]
nodejs can run (some) .ts files natively, but it won't load .ts files from node_modules, based largely on arguments from the typescript team that if node allowed .ts files in node_modules folks would publish .ts only packages to npm, which could cause performance issues for `tsc` as processsing .ts files is slower than .d.ts, and introduces nuance of tsc version and tsconfig settings.
@jakebailey @RyanCavanaugh 
I buy those arguments, but they convince me that typescript should implement _this_ feature - not type checking .ts files in node_modules - not that node should support type stripping in node_modules.

I setup a fresh tsc 5.9 project and wrote a module in node_modules/foo with .ts and .d.ts files, and typescript always preferred the .ts files instead of .d.ts files. TypeScript should have a recommended or default option to prefer .d.ts files, or ignore all .ts files in node_modules. That should fix the potential performance issue, and allow node to support type stripping in node_modules.
Folks can ship libraries with .ts to run and .d.ts files for type checks.
 


https://github.com/nodejs/typescript/issues/14#issuecomment-3903349663
https://github.com/nodejs/node/issues/57215#issuecomment-2689379347
